src/search/task_proxy.h

```diff
@@ -729,13 +718,90 @@ inline bool does_fire(const EffectProxy &effect, const State &state) {
     return true;
 }
 
-inline bool does_fire(const EffectProxy &effect, const GlobalState &state) {
-    for (FactProxy condition : effect.get_conditions()) {
-        FactPair condition_pair = condition.get_pair();
-        if (state[condition_pair.var] != condition_pair.value)
-            return false;
+inline bool State::operator==(const State &other) const {
+    assert(task == other.task);
+    if (registry != other.registry) {
+        std::cerr << "Comparing registered states with unregistered states "
+                  << "or registered states from different registries is "
+                  << "treated as an error because it is likely not "
+                  << "intentional."
+                  << std::endl;
+        utils::exit_with(utils::ExitCode::SEARCH_CRITICAL_ERROR);
+    }
+    if (registry) {
+        // Both states are registered and from the same registry.
+        return id == other.id;
+    }
+    // Both states are unregistered.
+    assert(values);
+    assert(other.values);
+    assert(!values->empty());
```

Comment:
Testing for emptiness seems odd here. If we want to test this, shouldn't we do so when creating states?